### PR TITLE
Release branch for 2.17.5

### LINKED ISSUE #8554: Crash using `isinstance` + `super` due to Astroid
### Bug description

When parsing the following ``a.py``:

```python
class A:
    def __init__(self):
        if not isinstance(super(), float):
            pass

```

there is a crash. Note however that this snippet is fine:

```python
class A:
    def __init__(self):
        if isinstance(super(), float):
            pass

```
with `not` removed. Also I believe it's nothing to do with `float`: you can replace this with a custom class, or another built-in.

### Command used

```shell
pylint a.py
```

### Pylint output

<details open>
    <summary>
        pylint crashed with a ``AstroidError`` and with the following stacktrace:
    </summary>

```python
Traceback (most recent call last):
  File "/home/bugreport/.venv/lib/python3.10/site-packages/astroid/inference_tip.py", line 33, in _inference_tip_cached
    result = _cache[func, node]
KeyError: (<function register_builtin_transform.<locals>._transform_wrapper at 0x7fbe7b0e5b40>, <Call l.3 at 0x7fbe7a87c910>)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 772, in _lint_file
    check_astroid_module(module)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1001, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1047, in _check_astroid_module
    walker.walk(node)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  [Previous line repeated 1 more time]
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/checkers/typecheck.py", line 1903, in visit_unaryop
    for error in node.type_errors():
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 4466, in type_errors
    return [
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 4466, in <listcomp>
    return [
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/inference.py", line 565, in _infer_unaryop
    for operand in self.operand.infer(context):
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 161, in infer
    results = list(self._explicit_inference(self, context, **kwargs))
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/inference_tip.py", line 40, in _inference_tip_cached
    result = _cache[func, node] = list(func(*args, **kwargs))
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/brain/brain_builtin_inference.py", line 175, in _transform_wrapper
    result = transform(node, context=context)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/brain/brain_builtin_inference.py", line 725, in infer_isinstance
    isinstance_bool = helpers.object_isinstance(obj_node, class_container, context)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/helpers.py", line 133, in object_isinstance
    obj_type = object_type(node, context)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/helpers.py", line 87, in object_type
    types = set(_object_type(node, context))
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/astroid/helpers.py", line 71, in _object_type
    raise AssertionError(f"We don't handle {type(inferred)} currently")
AssertionError: We don't handle <class 'astroid.objects.Super'> currently

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 736, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "/home/namc/pylintbugreport/.venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 774, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError

```

</details>

### Expected behaviour

No crash.

### Pylint version

```shell
pylint 3.0.0a6
astroid 2.15.1
Python 3.10.8 (main, Dec 19 2022, 16:27:32) [GCC 9.4.0]
```

### OS / Environment

linux (Linux)

### LINKED ISSUE #8598: Crash: AttributeError: 'Call' object has no attribute 'name'
### Bug description

Crash when parsing the following ``convo.py``: https://github.com/Dpeta/pesterchum-alt-servers/blob/62f3e21bd4699b04050b1d9ab3838aab9b2db28a/convo.py

Happens with ``2.17.2`` and ``3.0.0a6``.

### Command used

```shell
python -m pylint convo.py
```

### Pylint output

<details open>
    <summary>
        pylint crashed with a ``AstroidError`` and with the following stacktrace:
    </summary>

```python
Traceback (most recent call last):
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 772, in _lint_file
    check_astroid_module(module)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1001, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1047, in _check_astroid_module
    walker.walk(node)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/checkers/modified_iterating_checker.py", line 60, in visit_for
    self._modified_iterating_check_on_node_and_children(body_node, iter_obj)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/checkers/modified_iterating_checker.py", line 66, in _modified_iterating_check_on_node_and_children
    self._modified_iterating_check(body_node, iter_obj)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/checkers/modified_iterating_checker.py", line 96, in _modified_iterating_check
    obj_name = iter_obj.name
AttributeError: 'Call' object has no attribute 'name'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 736, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "/home/user/pchumenv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 774, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError

```

</details>

### Expected behavior

No crash.

### Pylint version

```shell
pylint 3.0.0a6
astroid 2.15.3
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201]
```

### OS / Environment

Arch Linux

### Additional dependencies
PyQt6

### LINKED ISSUE #8696: False positive `unused-variable` for `TYPE_CHECKING` imports
### Bug description

When `allow-global-unused-variables` is set to `false` in the configuration, the `unused-variable` warning is raised for imports made under `TYPE_CHECKING` guard clauses. This behavior seems inconsistent with `unused-import`, which is not raised for such cases.

```
# test.py
# pylint: disable=missing-module-docstring

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import math  # [unused-variable]
``` 

### Configuration

```ini
[tool.pylint.variables]
allow-global-unused-variables=false
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:6:4: W0612: Unused variable 'math' (unused-variable)
```

### Expected behavior

No warnings raised.

### Pylint version

```shell
pylint 3.0.0b1
astroid 3.0.0a2
Python 3.8.10 (default, Nov 14 2022, 12:59:47)
[GCC 9.4.0]
```

### OS / Environment

WSL2 Ubuntu 20.04 LTS

### Additional dependencies

_No response_

### LINKED ISSUE #8698: Crash "Fatal error while checking" with metaclass
### Bug description

When parsing the following file:

```python
"""Fatal error while checking."""
class Meta(type(object), type(object)):
    pass

class Class(metaclass=Meta):
    pass

```

pylint crashed with a ``AstroidError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\lint\pylinter.py", line 811, in _lint_file
    check_astroid_module(module)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\lint\pylinter.py", line 1085, in check_astroid_module
    retval = self._check_astroid_module(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\lint\pylinter.py", line 1135, in _check_astroid_module
    walker.walk(node)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\utils\ast_walker.py", line 94, in walk
    self.walk(child)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\utils\ast_walker.py", line 91, in walk
    callback(astroid)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\checkers\typecheck.py", line 1040, in visit_classdef
    if _is_invalid_metaclass(metaclass):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\checkers\typecheck.py", line 748, in _is_invalid_metaclass
    mro = metaclass.mro()
          ^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 3066, in mro
    return self._compute_mro(context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 3054, in _compute_mro
    unmerged_mro = list(clean_duplicates_mro(unmerged_mro, self, context))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 155, in clean_duplicates_mro
    raise DuplicateBasesError(
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (Meta), (type, object), (type, object), (type, type) for <ClassDef.Meta l.2 at 0x26ca8902bd0>.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\lint\pylinter.py", line 775, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\lint\pylinter.py", line 813, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError
```

### Configuration

_No response_

### Command used

```shell
pylint bug.py
```

### Pylint output

```shell
************* Module bug
bug.py:2:0: C0115: Missing class docstring (missing-class-docstring)
bug.py:2:0: E0241: Duplicate bases for class 'Meta' (duplicate-bases)
bug.py:5:0: C0115: Missing class docstring (missing-class-docstring)
Exception on node <ClassDef.Class l.5 at 0x26ca8903410> in file 'C:\Code\Bug\bug.py'
Traceback (most recent call last):
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\utils\ast_walker.py", line 91, in walk
    callback(astroid)
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\checkers\typecheck.py", line 1040, in visit_classdef
    if _is_invalid_metaclass(metaclass):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\pylint\checkers\typecheck.py", line 748, in _is_invalid_metaclass
    mro = metaclass.mro()
          ^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 3066, in mro
    return self._compute_mro(context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 3054, in _compute_mro
    unmerged_mro = list(clean_duplicates_mro(unmerged_mro, self, context))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bers\.pyenv-win-venv\envs\project_3.11\Lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 155, in clean_duplicates_mro
    raise DuplicateBasesError(
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (Meta), (type, object), (type, object), (type, type) for <ClassDef.Meta l.2 at 0x26ca8902bd0>.
bug.py:1:0: F0002: bug.py: Fatal error while checking 'bug.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\bers\AppData\Local\pylint\pylint\Cache\pylint-crash-2023-05-18-09-05-35.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 7.14/10, -7.14)
```

### Expected behavior

I have no idea if the code examples makes any sense, but `pylint` should not crash.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.5
Python 3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

Windows 10 21H2

### Additional dependencies

_No response_

### LINKED ISSUE #8719: False positive `unspecified-encoding` when passing encoding as kwarg
### Bug description

When passing an `encoding` as a kwarg to `open`, I get `unspecified-encoding` thrown.

```python
CSV_KWARGS = {"newline": "", "encoding": "utf-8"}
with open("foo.csv", **CSV_KWARGS):
    pass
```

### Configuration

_No response_

### Command used

```shell
pylint --score=false a.py
```

### Pylint output

```shell
************* Module quick_play
a.py:2:5: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
```

### Expected behavior

I think it would be cool if `pylint` wouldn't throw `unspecified-encoding` in this case.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.4
Python 3.10.10 (main, Feb 19 2023, 17:57:18) [Clang 14.0.0 (clang-1400.0.29.102)]
```

### OS / Environment

macOS Monterey v12.6

### Additional dependencies

_No response_

### LINKED ISSUE #8735: Checking incorrect usage of the keyword 'nonlocal'  results in fatal errors.
If a 'nonlocal'  variable are assigned with a variable outside a function, Pylint fails to handle it. See the following example.

example.py:
```python
nonlocal X
X = whatever
```

### The expected output:
There is no a fatal error.

### The actual output:
```
>> pylint example.py
************* Module pylint_example
Desktop/pylint_example.py:1:0: C0114: Missing module docstring (missing-module-docstring)
Desktop/pylint_example.py:3:0: E0117: nonlocal name X found without binding (nonlocal-without-binding)
Exception on node <Assign l.4 at 0x7fdb12c04a90> in file '/home/xxm/Desktop/pylint_example.py'
Traceback (most recent call last):
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 2043, in visit_assign
    self._check_self_cls_assign(node)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 2920, in _check_self_cls_assign
    scope = node.scope().parent.scope()
            ^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'scope'
Desktop/empirical_type_checker/bug_report/pylint_example.py:1:0: F0002: /home/xxm/Desktop/pylint_example.py: Fatal error while checking '/home/xxm/Desktop/pylint_example.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/xxm/.cache/pylint/pylint-crash-2023-06-06-15-56-31.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Test Environment:
Ubuntu 18.04,
CPython 3.11.3
Pylint: 3.0.0b1
astroid 3.0.0a3

### LINKED ISSUE #8740: The undefined variable '__all__' results in an inference failed
The undefined variable runners uses an undefined attributes \_\_all\_\_, pylint reports an inference failed. See the following example.

pylint_example.py
```
import sys
if sys.version_info >= (3, 7):
    __all__ += runners.__all__
```

The expected behavior:
Pylint does not crash.

The actual output:
```
>> pylint pylint_example.py
************* Module pylint_example
Desktop/pylint_example.py:19:0: C0303: Trailing whitespace (trailing-whitespace)
Desktop/pylint_example.py:56:0: C0304: Final newline missing (missing-final-newline)
Desktop/pylint_example.py:1:0: C0114: Missing module docstring (missing-module-docstring)
Desktop/pylint_example.py:55:4: E0602: Undefined variable '__all__' (undefined-variable)
Desktop/pylint_example.py:55:15: E0602: Undefined variable 'runners' (undefined-variable)
Exception on node <Module.pylint_example l.0 at 0x7f797324d190> in file '/home/xxm/Desktop/pylint_example.py'
Traceback (most recent call last):
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/utils/ast_walker.py", line 96, in walk
    callback(astroid)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 1364, in leave_module
    self._check_all(node, not_consumed)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 3049, in _check_all
    assigned = next(node.igetattr("__all__"))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xxm/.local/lib/python3.11/site-packages/astroid/bases.py", line 197, in _infer_stmts
    raise InferenceError(
astroid.exceptions.InferenceError: Inference failed for all members of [<AssignName.__all__ l.55 at 0x7f797324da10>].
Desktop/empirical_type_checker/bug_report/pylint_example.py:1:0: F0002: /home/xxm/Desktop/pylint_example.py: Fatal error while checking '/home/xxm/Desktop/pylint_example.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/xxm/.cache/pylint/pylint-crash-2023-06-07-00-29-01.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

#### Test Environment:
Ubuntu 18.04,
CPython 3.11.3
Pylint: 3.0.0b1
astroid 3.0.0a3

### LINKED ISSUE #8747: False positive for `inconsistent-return-statements` with method annotated with `NoReturn` return type
### Bug description

`repro.py`
```python
from typing import NoReturn

class Foo:
    def _never_returns() -> NoReturn:
        raise Exception

    def bar(self) -> int:
        try:
            return 1
        except Exception:
            self._never_returns()
```

### Configuration

```ini
Default Config
```

### Command used

```shell
pylint --disable=all --enable=inconsistent-return-statements repro.py
```

### Pylint output

```shell
************* Module repro
repro.py:7:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 0.00/10, +8.89)
```

### Expected behavior

```console
------------------------------------
Your code has been rated at 10.00/10
```

### Pylint version

```shell
My pylint configuration is pinned to 


pylint 2.15.8
astroid 2.13.5
Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]


But I was able to repro this on the latest released version of pylint (2.17.4 at time of writing).
```

### OS / Environment

```console
$ lsb_release -a
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.2 LTS
Release:	22.04
Codename:	jammy
$ uname -r
5.15.0-72-generic
```

### Additional dependencies

_No response_

---

I did an initial investigation, and _might_ have an idea of what's going on:

* During the `inconsistent-return-statements` lint, there's an attempt made infer whether the return type of a function call is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L2001-L2016

* But `_is_function_def_never_returning` only inspects the return type if the node passed in is a `nodes.FunctionDef`. But calling `node.func.inferred()[0]` on the method call (e.g. `self._never_returns()`) returns the corresponding method definition node as type `astroid.BoundMethod`. Since that isn't a `nodes.FunctionDef`, `_is_function_def_never_returning` never even looks to see if the return type is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L1956-L1957

### LINKED ISSUE #8754: Setting the identical name between an attribute and a variable  causes a fatal error
Setting the identical name between an attribute and a variable  causes a fatal error. See the following example.

pylint_example.py
```
attr = "test"
class T:
    T.attr = attr
```

### The expected behavior:

No fatal error

### Pylint output:
```
>> pylint pylint_example.py
************* Module pylint_example
pylint_example.py:82:0: C0103: Constant name "attr" doesn't conform to UPPER_CASE naming style (invalid-name)
pylint_example.py:83:0: C0115: Missing class docstring (missing-class-docstring)
pylint_example.py:83:0: C0103: Class name "T" doesn't conform to PascalCase naming style (invalid-name)
pylint_example.py:84:4: E0602: Undefined variable 'T' (undefined-variable)
Exception on node <Name.attr l.84 at 0x7f98c1d756d0> in file '/home/xxm/Desktop/pylint_example.py'
Traceback (most recent call last):
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 1616, in visit_name
    self._undefined_and_used_before_checker(node, stmt)
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 1657, in _undefined_and_used_before_checker
    action, nodes_to_consume = self._check_consumer(
                               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 1749, in _check_consumer
    found_nodes = current_consumer.get_next_to_consume(node)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xxm/.local/lib/python3.11/site-packages/pylint/checkers/variables.py", line 612, in get_next_to_consume
    if lhs.name == name:  # this name is defined in this very statement
       ^^^^^^^^
AttributeError: 'AssignAttr' object has no attribute 'name'
pylint_example.py:1:0: F0002: /home/xxm/Desktop/pylint_example.py: Fatal error while checking '/home/xxm/Desktop/pylint_example.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/xxm/.cache/pylint/pylint-crash-2023-06-09-15-25-11.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)

```
### Test environment:
pylint 3.0.0b1
astroid 3.0.0a5-dev0
Python 3.11.3
Ubuntu 18.04

### LINKED ISSUE #8774: no-argument call to copy.copy() crashes
### Bug description

```python
from copy import copy
copy()
```

### Configuration

```ini
n/a
```

### Command used

```shell
pylint a.py
```

### Pylint output

<details open>
    <summary>
        pylint crashed with a ``AstroidError`` and with the following stacktrace:
    </summary>

```
Traceback (most recent call last):
  File "/Users/user/pylint/pylint/lint/pylinter.py", line 788, in _lint_file
    check_astroid_module(module)
  File "/Users/user/pylint/pylint/lint/pylinter.py", line 1017, in check_astroid_module
    retval = self._check_astroid_module(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/pylint/pylint/lint/pylinter.py", line 1063, in _check_astroid_module
    walker.walk(node)
  File "/Users/user/pylint/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/Users/user/pylint/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/Users/user/pylint/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/Users/user/pylint/pylint/checkers/stdlib.py", line 563, in visit_call
    self._check_shallow_copy_environ(node)
  File "/Users/user/pylint/pylint/checkers/stdlib.py", line 514, in _check_shallow_copy_environ
    arg = utils.get_argument_from_call(node, position=0)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/pylint/pylint/checkers/utils.py", line 736, in get_argument_from_call
    raise NoSuchArgumentError
pylint.checkers.utils.NoSuchArgumentError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/user/pylint/pylint/lint/pylinter.py", line 752, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "/Users/user/pylint/pylint/lint/pylinter.py", line 790, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError
```

</details>

### Expected behavior

no crash

### Pylint version

```shell
pylint 3.0.0b1
astroid 3.0.0a6-dev0
Python 3.11.2
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### LINKED ISSUE #8830: False positive `invalid-enum-extension` for Enum parent class with annotations
### Bug description

Pylint incorrectly flags an Enum inherited from a parent Enum which has annotations. This is useful when overriding `__new__` or `__init__` on the parent Enum to add extra data to the Enum members while also keeping the codebase typed. Example:

`example.py`
```
from enum import Enum


class ColorEnum(Enum):
    red: int
    green: int
    blue: int

    def __init__(self, red: int, green: int, blue: int) -> None:
        self.red = red
        self.green = green
        self.blue = blue

class Pastel(ColorEnum):
    SAGE = (170, 200, 167)
```

### Configuration

_No response_

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:14:0: E0244: Extending inherited Enum class "ColorEnum" (invalid-enum-extension)
```

### Expected behavior

No output

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.5
Python 3.8.16 (default, Mar 15 2023, 13:36:46)
[Clang 14.0.0 (clang-1400.0.29.202)]
```

### OS / Environment

macOS Ventura 13.2.1

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
What's new in Pylint 2.17.5?
----------------------------
Release date: 2023-07-26

False Positives Fixed
---------------------

- Fix a false positive for ``unused-variable`` when there is an import in a
  ``if TYPE_CHECKING:`` block and ``allow-global-unused-variables`` is set to
  ``no`` in the configuration.

  Closes #8696 

- Fix false positives generated when supplying arguments as ``**kwargs`` to IO
  calls like open().

  Closes #8719

- Fix a false positive where pylint was ignoring method calls annotated as
  ``NoReturn`` during the ``inconsistent-return-statements`` check.

  Closes #8747

- Exempt parents with only type annotations from the ``invalid-enum-extension``
  message.

  Closes #8830

Other Bug Fixes
---------------

- Fixed crash when a call to ``super()`` was placed after an operator (e.g.
  ``not``).

  Closes #8554 

- Fix crash for ``modified-while-iterating`` checker when deleting
  members of a dict returned from a call.

  Closes #8598 

- Fix crash in ``invalid-metaclass`` check when a metaclass had duplicate
  bases.

  Closes #8698 

- Avoid ``consider-using-f-string`` on modulos with brackets in template.

  Closes #8720. 

- Fix a crash when ``__all__`` exists but cannot be inferred.

  Closes #8740

- Fix crash when a variable is assigned to a class attribute of identical name.

  Closes #8754 

- Fixed a crash when calling ``copy.copy()`` without arguments.

  Closes #8774 

Other Changes
-------------

- Fix a crash when a ``nonlocal`` is defined at module-level.

  Closes #8735

### CODE REVIEW

**Review by mbyrnepr2 — APPROVED** [review]
👍

**Comment by Pierre-Sassoulas** [comment]
We won't merge this back in the main branch because 3.0 is too far away from 2.X at this point. This would break the pre-commit autoupdate, so we might do a history only merge.

File: doc/user_guide/configuration/all-options.rst [hunk 1]
```diff
@@ -233,7 +233,7 @@ Standard Checkers
 
    confidence = ["HIGH", "CONTROL_FLOW", "INFERENCE", "INFERENCE_FAILURE", "UNDEFINED"]
 
-   disable = ["raw-checker-failed", "bad-inline-option", "locally-disabled", "file-ignored", "suppressed-message", "useless-suppression", "deprecated-pragma", "use-symbolic-message-instead", "consider-using-augmented-assign"]
+   disable = ["consider-using-augmented-assign"]
```
Comment (jacobtylerwalls) [inline doc/user_guide/configuration/all-options.rst:236]: Quick question: from where did this change arise?

When I `--generate-toml-config` on 2.17.5 I get:

```
disable = ["raw-checker-failed", "bad-inline-option", "locally-disabled", "file-ignored", "suppressed-message", "deprecated-pragma", "consider-using-augmented-assign", "attribute-defined-outside-init", "invalid-name", "missing-module-docstring", "missing-class-docstring", "missing-function-docstring", "protected-access", "too-few-public-methods", "line-too-long", "too-many-lines", "trailing-whitespace", "missing-final-newline", "trailing-newlines", "bad-indentation", "unnecessary-semicolon", "multiple-statements", "superfluous-parens", "mixed-line-endings", "unexpected-line-ending-format", "fixme"]
```

Comment (Pierre-Sassoulas) [reply doc/user_guide/configuration/all-options.rst:236]: Applied ``tox -e docs`` (probably from tbump)

Comment (jacobtylerwalls) [reply doc/user_guide/configuration/all-options.rst:236]: Interesting. So there's a bug somewhere here, then?

Comment (Pierre-Sassoulas) [reply doc/user_guide/configuration/all-options.rst:236]: Maybe it's based on the local pylintrc ?

**Pull request merged**